### DISCUSSION TITLE
HyperHDR Kodi addon and suspend

### DISCUSSION THREAD

**Discussion by sjl68** [awawa-dev/HyperHDR#740] [Q&A] [answered]
I have Libreelec installed on a mini pc and I am using the new HyperHDR addon (works brilliantly thnak you Awawa). Additonally I am using HomeAssistant to control certain aspects of the lighting.
 
At the moment HA turns on WLED 30 minutes before sunset which then triggers some other lights to illuminate. 

When I start the Librelec pc HyperHDR functions as expected. If I then choose to suspend the LE pc HyperHDR doesnt release control so the leds will stay in a dimmed state depending on what screen was displayed when I chose to suspend the pc.

![image](https://github.com/awawa-dev/HyperHDR/assets/90412811/8e5ed1d9-7d11-4ed7-b5e9-f3d04f82bb38)

I have tried to add a trigger to HA with no success (noob to HA) and I do not have the skills to write a script for Libreelec to stop the addon on suspend.

Can anyone suggest a way forward please.

**Comment by awawa-dev** [comment]
Hi
On Linux, HyperHDR detects the sleep/wake up events using DBus. I'm not sure if DBus works on Libreelec. Need full logs that capture the moment you put your PC running Libreelec to sleep and then wake it up. If this event is not detected then WLED will not be turned off nicely.

Alternatively, you can use 'udpraw' instead of the 'wled' driver in HyperHDR: it is compatible with WLED and, due to its simplicity, generally behaves better in the event of an unexpected communication break.

**Reply by sjl68** [reply]
Thank you, I have tried udpraw just now but the leds wil not illuminate. I will do a restart of everything tomorrow as it is 12.20 am local time and I am having a glass of whiskey or two so not really giving this my full attention :).

I will provide logs using WLED tomorrow too.

**Reply by awawa-dev** [reply]
'udpraw' for 99.9% should work, many users use it  (unless maybe you are testing latest WLED that has some stability problems especially on esp8266?) Maybe it's some WLED configuration issue? udpraw wont enable for you anything in WLED, if the LEDs are disabled or brightness is set to 0, it won't override it in WLED contrary to WLED driver.

**Reply by sjl68** [reply]
Here is the log from a fresh boot and a suspend and WOL event.

LEDs are not set to 0 brightness nor disabled in WLED either - just tried UDPRaw again and no difference - logs show eveyhting is happening as expected but the LED's dont react.

Thanks again.

<logs>

**Reply by awawa-dev** [reply]
1 You use 'override brightness' in your HyperHDR WLED driver settings. If you disable it, does it still works?
2 No suspend/resume events were detected in the logs

**Reply by awawa-dev** [reply]
I've completely forgot about it, but DBus event detection was fixed recently: https://github.com/awawa-dev/HyperHDR/commit/83c1f62d2dfb85b5f81fd0c057ddf00490e0def9 v20beta2 should be available soon

**Reply by sjl68** [reply]
I just tried disabling 'overide brightness' and it made no difference to the suspend behaviour.

I will keep an eye out for v20beta2 and hope that it is the solution. Great work!!

**Reply by awawa-dev** [reply]
Could you uninstall you current version first and install pre.v20beta2.generic_PC.hyperhdr.zip from https://github.com/awawa-dev/LibreELEC.tv/releases/tag/v20beta1 ?

**Reply by sjl68** [reply]
OK I unistalled as advised and added the new zip file - I cannot see any difference I am afraid - tried udpraw and that didnt work either. By the way I did as a test start from scratch and did the config from fresh after the new install.

On a positive note the toggle for black bar dectection in remote control now displays the fact that I am indeed using that and it also displays correctly on the main page.

**Reply by awawa-dev** [reply]
Did you check the HyperHDR logs for messages like "The system is going to sleep" or "The system is going to wake up" after you suspend/resume the system?

**Reply by sjl68** [reply]
Just tried it again and have added the log for you to decode :)

**Reply by awawa-dev** [reply]
Thank you for the logs and support :) It seems that the suspend/resume events are still not detected, this time maybe because they are not submitted in the DBus service org.freedesktop.login1. I will install Libreelec tomorrow to test it further.

**Reply by awawa-dev** [reply]
Stupid me: my virtual machine that I use to build HyperHDR plugins for Libreelec had DBus support disabled. And as a result, the addon also didnt support DBus and no wonder it didn't work. I have a huge request for you to test the latest version, because it is the first version for Libreelec that can work with DBus at all (if Libreelec supports DBus for power management). I changed the name a bit to pre.v20beta2.Generic_PC.hyperhdr.**dbus**.zip Uninstall the old one and install this: you do not need to configure HyperHDR from scratch.

**Reply by sjl68** [reply]
I have installed the new file, the behaviour has changed slightly in that the LED's dont dim  now on suspend they stay on the brightness and colour of the screen I was on when I chose suspend.

**Reply by awawa-dev** [reply]
Well, something has changed so probably it's a good sign. Did "The system is going to sleep" or "The system is going to wake up" appear in HyperHDR logs during suspend/resume events? You can post logs here, I will delete them later (or send me email).

**Reply by sjl68** [reply]
Heres a log from this morning the time stamp 09.52:38 is when I selected suspend. Theres nothng after that mentioning sleep or wake up.

...

**Reply by sjl68** [reply]
When I was looking for solutions I found an old post on the LE forum that referenced Suspend and Resume scripts - seems the info has been removed from the LE Wiki. But would something like that work?

Kill HyperHDR then restart after resume? I dont know the code or command required unfortunately.

#!/bin/sh
case "$1" in
   pre)
     <do something on suspend>
          ;;
   post)
     <do something on resume>
          ;;
esac

https://forum.libreelec.tv/thread/15470-suspend-and-resume-scripts-wiki-incorrect/

**Reply by awawa-dev** [reply]
I have an idea: could you enable ssh (libreelec settings and you should see the ssh options), use ssh(ex. putty) to connect to libreelec and check if your OS has 'curl' or 'wget' by executing for example `/usr/bin/curl`

**Reply by sjl68** [reply]
I hope this helps :)

LibreELEC (official): 11.0.4 (Generic-legacy.x86_64)
LibreELEC:~ # /usr/bin/curl
curl: try 'curl --help' for more information
LibreELEC:~ # curl --help
Usage: curl [options...] <url>
 -d, --data <data>          HTTP POST data
 -f, --fail                 Fail fast with no output on HTTP errors
 -h, --help <category>      Get help for commands
 -i, --include              Include protocol response headers in the output
 -o, --output <file>        Write to file instead of stdout
 -O, --remote-name          Write output to a file named as the remote file
 -s, --silent               Silent mode
 -T, --upload-file <file>   Transfer local FILE to destination
 -u, --user <user:password> Server user and password
 -A, --user-agent <name>    Send User-Agent <name> to server
 -v, --verbose              Make the operation more talkative
 -V, --version              Show version number and quit

This is not the full help, this menu is stripped into categories.
Use "--help category" to get an overview of all categories.
For all options use the manual or "--help all".
LibreELEC:~ # /usr/bin/wget
BusyBox v1.32.1 (2023-12-20 00:08:10 UTC) multi-call binary.

Usage: wget [-c|--continue] [--spider] [-q|--quiet] [-O|--output-document FILE]
        [-o|--output-file FILE] [--header 'header: value'] [-Y|--proxy on/off]
        [--no-check-certificate]
        [-P DIR] [-S|--server-response] [-U|--user-agent AGENT] [-T SEC] URL...

Retrieve files via HTTP or FTP

        --spider        Only check URL existence: $? is 0 if exists
        --no-check-certificate  Don't validate the server's certificate
        -c              Continue retrieval of aborted transfer
        -q              Quiet
        -P DIR          Save to DIR (default .)
        -S              Show server response
        -T SEC          Network read timeout is SEC seconds
        -O FILE         Save to FILE ('-' for stdout)
        -o FILE         Log messages to FILE
        -U STR          Use STR for User-Agent header
        -Y on/off       Use proxy

**Reply by awawa-dev** [reply, marked as answer]
Thank you! New updated release pre.v20beta2.Generic_PC.hyperhdr.**service**.zip
You should restart Libreelec after you install it. What should be found in the logs this time after you suspend/resume the OS is "COMP_ALL"

**Reply by awawa-dev** [reply]
It should automatically invoke two commands in the OS during Suspend/Resume events 
Suspend:
`/usr/bin/curl http://127.0.0.1:8090/json-rpc?request=%7B%22command%22:%22componentstate%22,%22componentstate%22:%7B%22component%22:%22ALL%22,%22state%22:false%7D%7D`
Resume:
`/usr/bin/curl http://127.0.0.1:8090/json-rpc?request=%7B%22command%22:%22componentstate%22,%22componentstate%22:%7B%22component%22:%22ALL%22,%22state%22:true%7D%7D`

**Reply by sjl68** [reply]
You sir are a genius!! That works now - thats fantastic thank you very much. The time and effort you put into this project never ceases to amaze me!

I have only snipped part of the log to save filling this up with all the uneeded bits before I hit suspend. 
...

**Reply by awawa-dev** [reply]
This is great news and we finally made it! Thank you for your help in testing development versions: without it the problem would not have been solved. The next version of the add-on for Libreelec (v20beta2) will have this solution implemented as standard.

**Reply by sjl68** [reply]
I have two other issues but need to do some more testing as one of them might be my equipment causing one issue :).

**Reply by awawa-dev** [reply]
OK, let me know later what you find out. Now is a good time for testing before the final version.

**Comment by sjl68** [comment]
Hi Awawa,

Sorry to bring this topic back up again but with the new Kodi add on the suspend function has broken and I get the same fault that I previously reported.

I currently don’t have any logs but happy to grab them later for you.

**Comment by awawa-dev** [comment]
Maybe they forgot to include my power management script? https://github.com/awawa-dev/LibreELEC.tv/blob/libreelec-12.0-hyperhdr/packages/addons/service/hyperhdr/source/system.d/service.hyperhdr-power-mng.service

**Reply by sjl68** [reply]
Ok thanks can I add that script in or does it have to be done by the addon maintainers? I assume the Libreelec team?

**Reply by awawa-dev** [reply]
Yes, they currently maintain HyperHDR Kodi addon. Or you can add this script yourself to your setup but you must repeat this procedure each time when you reinstall Kodi from scratch (and probably after upgrade Kodi to next major version).

**Reply by sjl68** [reply]
I have posted on the LE forum, I tried to add the script but it doesn’t work with the latest releases as my post on there describes 

https://forum.libreelec.tv/thread/29780-hyperhdr-suspend-not-working/
